sdk/core/Azure.Core/src/Shared/OperationInternalOfT.cs

```diff
@@ -223,27 +247,53 @@ public Response<T> WaitForCompletion(TimeSpan pollingInterval, CancellationToken
             return Response.FromValue(Value, rawResponse);
         }
 
-        /// <summary>
-        /// Sets the <see cref="OperationInternal{T}"/> state immediately.
-        /// </summary>
-        /// <param name="state">The <see cref="OperationState{T}"/> used to set <see cref="OperationInternalBase.HasCompleted"/> and other members.</param>
-        public void SetState(OperationState<T> state)
+        protected override async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToken cancellationToken)
         {
-            if (state.HasCompleted && state.HasSucceeded)
+            using var asyncLock = await _stateLock.GetLockOrValueAsync(async, cancellationToken).ConfigureAwait(false);
```

Comment:
Let's add a comment here to explain the difference between this lock and the lock we keep for the value.

Reply:
We can't, but there is no reason to have multiple loops requesting server status - one in enough.